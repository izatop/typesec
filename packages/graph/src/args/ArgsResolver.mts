import {assert} from "@typesec/core";
import {object, type KeyOf, type Promisify} from "@typesec/the";
import type {Args, Graph, Node, Proto} from "../interfaces.mts";

export class ArgsResolver<A extends Args<any>> {
    readonly #node: Proto.ArgsNode<any, A>;

    constructor(node: Proto.ArgsNode<any, A>) {
        this.#node = node;
    }

    public async resolve(payload: unknown): Promise<Graph.Extract<A>> {
        assert(object.isObject(payload), `ARG_RESOLVER_WRONG_PAYLOAD`);

        const result = {} as Graph.Extract<A>;
        for (const key of this.#node.keys()) {
            const member = this.#node.members[key];
            const value = await this.#resolveType(member, payload[key]);
            Object.assign(result, {[key]: value});
        }

        return result;
    }

    async #resolveType<K extends KeyOf<A, string>>({type}: Node.MakeArgs<A>[K], payload: unknown): Promise<any> {
        switch (type.kind) {
            case "args":
                return this.#resolveArgs(type, payload);

            case "complex":
                return this.#resolveComplex(type, payload);

            case "null":
            case "array":
                return this.#resolveComposite(type, payload);

            case "primitive":
                return this.#resolvePrimitive(type, payload);
        }
    }

    async #resolvePrimitive<T extends Graph.Primitive>(type: Proto.Primitive<any, T>, payload: unknown): Promise<T> {
        assert(type.validate(payload), `Wrong ${type.name} primitive`);

        return payload;
    }

    async #resolveComposite<P extends Proto.Composite<any, any, any>>(
        type: P,
        payload: unknown,
    ): Promise<Graph.Extract<A>> {
        assert(type.validate(payload), `Wrong composite value`);

        return payload;
    }

    async #resolveArgs<T extends Proto.ArgsNode<any, any>>(type: T, payload: unknown): Promise<Graph.Extract<T>> {
        const resolver = new ArgsResolver(type);

        const result = await resolver.resolve(payload);
        assert(type.validate(result), `${type.name} Validation failed`);

        return result;
    }

    #resolveComplex<T>(type: Proto.Complex<any, T, any>, payload: unknown): Promisify<T> {
        assert(Array.isArray(payload), `Wrong format`);
        const [id, serialized] = payload;

        assert(id === type.name, `Wrong format identifier`);
        assert(type.format.validate(serialized), `Wrong serialized data`);

        return type.deserialize(serialized);
    }
}
