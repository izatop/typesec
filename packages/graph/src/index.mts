export * from "./assert.mjs";
export * from "./contextual.mjs";
export * from "./graph.mjs";
export * from "./interfaces.mjs";
export * from "./proto.mjs";
export * from "./query.mjs";
export * from "./scalars.mjs";
export * from "./scope.mjs";
