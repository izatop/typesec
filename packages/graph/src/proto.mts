import {has, is, isNullable, isObject} from "@typesec/the";
import type {Graph, Proto} from "./interfaces.mts";

export const ProtoSymbol = Symbol();

export function proto<ID extends string, T, O extends Graph.Primitive>(
    args: Proto.Complex<ID, T, O>,
): Proto.Factory<Proto.Complex<ID, T, O>>;
export function proto<ID extends string, T extends Graph.Primitive, K extends Proto.Kind>(
    args: Proto.Type<ID, T, K>,
): Proto.Factory<Proto.Type<ID, T, K>>;
export function proto(args: Proto.Any): Proto.Factory<any> {
    const factory = () => args;

    Object.assign(factory, {type: args});

    return factory as Proto.Factory<any>;
}

export function primitive<ID extends string, T extends Graph.Primitive>(
    args: Omit<Proto.Type<ID, T, any>, "kind">,
): Proto.Factory<Proto.Primitive<ID, T>> {
    return proto({
        kind: "primitive",
        ...args,
    });
}

export function complex<ID extends string, T, O extends Graph.Primitive>(
    args: Omit<Proto.Complex<ID, T, O>, "kind">,
): Proto.Factory<Proto.Complex<ID, T, O>> {
    return proto({
        kind: "complex",
        ...args,
    });
}

export function nullable<C extends Proto.Any>(child: C): Proto.NullishType<C> {
    return {
        kind: "null",
        name: `nullable(${child.name})`,
        validate: (value) => isNullable(value) || child.validate(value),
        child,
    };
}

export function array<C extends Proto.Any>(child: C): Proto.ArrayType<C> {
    return {
        kind: "array",
        name: `array(${child.name})`,
        validate: (value): value is Proto.Infer<C> => Array.isArray(value) && value.every((v) => child.validate(v)),
        child,
    };
}

export function getProtoOf<T extends Proto.Any>(factory: Proto.Factory<T>): T {
    return factory.type;
}

export function isProto(value: unknown): value is Proto.Any {
    return isObject(value) && has(value, "name", "validate") && is(value.validate, "function");
}
