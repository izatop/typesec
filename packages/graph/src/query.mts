import {type IsNever, type Rec} from "@typesec/the";
import type {Graph, Query} from "./interfaces.mts";
import {Scope, type ScopeQueryArgs} from "./scope/Scope.mts";

export type QueryArgs<G extends Graph<any>, Q extends Query<G>, S extends Rec, C> =
    IsNever<C> extends false ? [scope: Scope<G, S, C>, query: Q, context: C] : [scope: Scope<G, S, C>, query: Q];

export async function query<G extends Graph<any>, Q extends Query<G>, S extends Rec, C>(
    ...[scope, ...args]: QueryArgs<G, Q, S, C>
): Promise<Query.Result<G, Q>> {
    return scope.query(...(args as ScopeQueryArgs<Q, C>));
}
