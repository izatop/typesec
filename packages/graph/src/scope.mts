import {type Rec} from "@typesec/the";
import type {Graph, Node, Proto} from "./interfaces.mts";
import {Scope} from "./scope/Scope.mts";

export function scope<G extends Graph<any>, S extends Rec, C = never>(
    node: Proto.GraphNode<any, G, S, C>,
    resolve: Node.ScopeResolve<S, C>,
): Scope<G, S, C> {
    return new Scope(node, resolve);
}
