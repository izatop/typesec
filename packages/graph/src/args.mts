import {assert} from "@typesec/core";
import {isObject, object} from "@typesec/the";
import type {Args, Graph, Node, Proto} from "./interfaces.mts";
import {isProto} from "./proto.mts";

export function args<A extends Args<any>, ID extends string = string>(
    name: ID,
    members: Node.MakeArgs<A>,
): Proto.ArgsNode<ID, A> {
    return {
        name,
        members,
        kind: "args",
        keys: () => object.keys(members),
        has: (key) => object.hasKeyOf(members, key),
        validate: (value): value is Graph.Infer<A> => {
            const entries = Object.entries(members);
            if (!isObject(value)) {
                return false;
            }

            for (const [key, member] of entries) {
                assert(isProto(member), "@TODO isMember(member)");
                assert(member.validate(value[key]), "@TODO member.validate(value)");
            }

            return true;
        },
    };
}
