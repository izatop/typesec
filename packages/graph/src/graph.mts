import {assert} from "@typesec/core";
import {isObject, object, toEntries, type Rec} from "@typesec/the";
import type {Graph, Node, Proto} from "./interfaces.mts";
import {isProto} from "./proto.mts";

export function graph<G extends Graph<any>, S extends Rec, ID extends string = string, C = never>(
    name: ID,
    members: Node.Make<G, S, C>,
): Proto.GraphNode<ID, G, S, C> {
    return {
        name,
        members,
        kind: "graph",
        keys: () => object.keys(members),
        has: (key) => object.hasKeyOf(members, key),
        validate: (value): value is Graph.Infer<G> => {
            const entries = toEntries(members);
            if (!isObject(value)) {
                return false;
            }

            for (const [key, {type}] of entries) {
                assert(isProto(type), `@TODO [${name}.${key}] isProto(member)`);
                assert(type.validate(value[key]), `@TODO [${name}.${key}] validate(${value[key]})`);
            }

            return true;
        },
    };
}
