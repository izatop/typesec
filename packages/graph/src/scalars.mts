import {complex, primitive} from "./proto.mts";

export const string = primitive({
    name: "string",
    validate: (value): value is string => typeof value === "string",
});

export const boolean = primitive({
    name: "bool",
    validate: (value): value is boolean => typeof value === "boolean",
});

export const float = primitive({
    name: "float",
    validate: (value): value is number => typeof value === "number",
});

export const int = primitive({
    name: "int",
    validate: (value): value is number => typeof value === "number" && value % 1 === 0,
});

export const bigint = complex({
    name: "bigint",
    format: string(),
    validate: (value): value is bigint => typeof value === "bigint",
    deserialize: (value) => BigInt(value),
    serialize: (value) => value.toString(),
});

export const datetime = complex({
    name: "datetime",
    format: string(),
    validate: (value): value is Date => value instanceof Date && value.toString() !== "Invalid Date",
    deserialize: (value) => new Date(value),
    serialize: (value) => value.toISOString(),
});

export const scalars = {
    string,
    boolean,
    float,
    int,
    bigint,
    datetime,
};
