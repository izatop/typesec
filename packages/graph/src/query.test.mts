import {describe, expect, test} from "bun:test";
import {args} from "./args.mts";
import {graph} from "./graph.mts";
import type {Args, Graph, Node} from "./interfaces.mts";
import {array} from "./proto.mts";
import {query} from "./query.mts";
import {scalars} from "./scalars.mts";
import {scope} from "./scope.mts";

describe("query()", () => {
    test("should resolve primitives", async () => {
        type S = {i: number; f: number; s: string; b: boolean};
        type G = Graph<S>;
        const Node = graph<G, S>("Test", {
            i: scalars.int,
            f: scalars.float,
            s: scalars.string,
            b: scalars.boolean,
        });

        const entry = scope(Node, () => ({i: 1, f: 1.1, b: true, s: "s"}));
        const res = await query(entry, {
            i: true,
            f: true,
            b: true,
            s: true,
        });

        expect(res).toMatchSnapshot();
    });

    test("should resolve virtual nodes", async () => {
        type G = Graph<{random: number}>;
        const value = Math.random();
        const Node = graph<G, {}>("T", {
            random: {type: scalars.float(), resolve: () => value},
        });

        const entry = scope(Node, () => ({id: 1}));
        const res = await query(entry, {random: true});
        expect(res.random).toBe(value);
    });

    test("should resolve complex type", async () => {
        const value = new Date();

        type S = {date: Date};
        type G = Graph<{date: Node<Date>}>;
        const Node = graph<G, S>("T", {
            date: {type: scalars.datetime()},
        });

        const entry = scope(Node, () => ({date: value}));
        const res = await query(entry, {date: true});
        expect(res.date).toBe(value);
    });

    test("should support args", async () => {
        type A = Args<{x: number[]}>;
        type G = Graph<{adds: Node<number, A>}>;
        const NodeArgs = args<A>("A", {x: {type: array(scalars.int())}});
        const Node = graph<G, {}>("T", {
            adds: {
                args: NodeArgs,
                type: scalars.int(),
                resolve: ({args}) => args.x.reduce((l, r) => l + r, 0),
            },
        });

        const entry = scope(Node, () => ({n1: 1}));
        const res = await query(entry, {adds: [{x: [1, 2, 3]}]});
        expect(res.adds).toEqual(1 + 2 + 3);
    });

    test("should query in scope", async () => {
        type S = {n1: number};
        type G = Graph<{n1: number; n2: number}>;
        const Node = graph<G, S>("T", {
            n1: {type: scalars.int()},
            n2: {
                type: scalars.int(),
                resolve: async ({subquery}) => {
                    const {n1} = await subquery.query({n1: 1});

                    return n1 + 1;
                },
            },
        });

        const entry = scope(Node, () => ({n1: 1}));
        const res = await query(entry, {n1: true, n2: true});
        expect(res).toEqual({n1: 1, n2: 2});
    });

    test("should run subquery", async () => {
        type S = {n1: number};
        type G = Graph<{n1: number; n2: number}>;
        const Node = graph<G, S>("T", {
            n1: {type: scalars.int()},
            n2: {
                type: scalars.int(),
                resolve: async ({subquery}) => {
                    const {n1} = await subquery.query({n1: 1});

                    return n1 + 1;
                },
            },
        });

        const entry = scope(Node, () => ({n1: 1}));
        const res = await query(entry, {n1: true, n2: true});
        expect(res).toEqual({n1: 1, n2: 2});
    });

    test("should run deep query", async () => {
        const id = 1;
        const name = "B. Node";

        type A = {};
        type B = {id: number; name: string};
        type GA = Graph<{child: Node<GB>; children: Node<GB[]>}>;
        type GB = Graph<B>;
        const NB = graph<GB, B>("B", {id: scalars.int, name: scalars.string});
        const NA = graph<GA, A>("A", {
            child: {
                type: NB,
                resolve: () => {
                    return {id, name};
                },
            },
            children: {
                type: array(NB),
                resolve: () => [{id, name}],
            },
        });

        const s = scope(NA, {});

        const res = await s.query({child: [{name: 1}], children: [{id: 1}]});
        expect(res).toEqual({child: {name}, children: [{id}]});

        const aliased = await s.query({child: {a: [{name: 1}], b: [{id: 1}]}});
        expect(aliased).toEqual({child: {a: {name}, b: {id}}});
    });
});
