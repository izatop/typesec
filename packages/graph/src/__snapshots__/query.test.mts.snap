// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`query() should resolve primitives 1`] = `
{
  "b": true,
  "f": 1.1,
  "i": 1,
  "s": "s",
}
`;
